### ISSUE TITLE
Properties may not be instanced at OnValidate or Awake

### ISSUE THREAD

**Issue by smmathews** [jacobdufault/fullinspector#107]
we have an issue where instanced properties will sometimes be null in OnValidate or Awake, when they are later filled in with the data the designer has set in the editor. How would you suggest validating this data for errors?

```c#
[Serializable]
public abstract class LocationData  {
    public abstract CheckData(BaseBehavior behavior);
}
[Serializable]
public class CityLocationData : LocationData {
    public BuildingData building { get; private set;}
    public override CheckData(BaseBehavior behavior) {
        if(building == null)
            building.LogError("no building data has been specified",behavior);
}
public class MapPoint : BaseBehavior<FullInspector.FullSerializerSerializer> {
    public LocationData locationData;
    public void OnValidate() {
        if(locationData == null)
            Debug.LogError("locationData has not been instantiated",this);
        else
            locationData.CheckData(this);
    }
}
```

**Comment by jacobdufault** [comment]
For `MapPoint.locationData`, you can set `fiSettings.AutomaticReferenceInstantation` to true, though I do not recommend this. With this setting set to true, then when an object is restored Full Inspector will go through every field/property and attempt to instantiate it so it is not null, which will cause a negative performance impact.

For the data validation, I would suggest just handling the scenario where the value can be null. If it is null, then just allocate an instance or log a warning message.

**Comment by smmathews** [comment]
Sorry, you're missing the point. these properties have already been set by the designer on the MapPoint instances throughout the map, so I can't just make another instance (it won't have the designer's data) so if I can't check the designer's work in OnValidate, where would you suggest I validate the data?

**Comment by jacobdufault** [comment]
Maybe I'm still misunderstanding - if so, can you clarify some more? The designer has specified the data and when you go to validate it it is set to null - as in, it isn't getting deserialized? If this is the case, can you send me more specific repro instructions?

Otherwise, can you just skip out validating if the property is set to null? If it is null then we know that the designer has not looked at the object.

The best we can do beyond `null` is something like `Activator.CreateInstance` or `FormatterServices.GetUninitializedObject` which either invokes the default constructor or returns an object with zeroed out memory. `fiSettings.AutomaticReferenceInstantation` gives you this behavior instead of null.

Thanks

**Comment by smmathews** [comment]
Download this project I just whipped up to demonstrate the issue
https://drive.google.com/file/d/0B5vFFD3Xr4hZOFVKcTA3RUZyVjA/view?usp=sharing
Open the Scene file
Import Full Inspector (left out of this project so I don't give away your source code)
Run the game. Stop the game. Run the game again. Notice that even though "Sears Tower" and "Airport" have their Location Data set, OnValidate never sees it as anything except null, so the same two errors keep popping up.

**Comment by jacobdufault** [comment]
Thanks! I think I'm able to reproduce this. I'm amazed this hasn't been caught yet, as the key issue seems to be that deserialization is not running before `OnValidate()` is called.

One easy fix for this is to simply call `RestoreState()` in `OnValidate`, ie,

```c#
public class MapPoint : BaseBehavior<FullInspector.FullSerializerSerializer> {
    public LocationData locationData;
    public void OnValidate() {
        RestoreState();

        if(locationData == null)
            Debug.LogError("locationData has not been instantiated",this);
        else
            locationData.CheckData(this);
    }
}
```

(If this does not actually fix the issue please let me know - I might be doing the repro wrong).

This is obviously not optimal. I'll try to think of a better solution so that it just works - the worst case being making `OnValidate()` a virtual method call.

**Comment by jacobdufault** [comment]
This is fixed in 2.6. Let me know if you want a preview build that includes the fix.

Note: to use the fix, `OnValidate` is now a virtual method. You need to override it and make sure to call `base.OnValidate()`.
